routing/message/handlers/passthrough/device_handler.go

```diff
@@ -0,0 +1,305 @@
+	}
+
+	h.Debug("Send message", map[string]interface{}{"topic": topic, "payload": string(payload)})
+
+	message := message.NewMessage(watermill.NewUUID(), payload)
+	message.SetContext(connector.SetTopicToCtx(message.Context(), topic))
+	return message
+}
+
+// getFeatureProperties return all feature properties, including the definition.
+func (h *deviceHandler) getFeatureProperties(obj interface{}) (interface{}, bool) {
+	res := map[string]interface{}{}
+	if feature, ok := obj.(map[string]interface{}); ok {
+		if pObj, ok := feature[valuePropertiesTag]; ok {
+			if properties, ok := pObj.(map[string]interface{}); ok {
+				res = properties
+			}
+		}
+
+		if dObj, ok := feature[valueDefinitionTag]; ok {
+			res[valueDefinitionTag] = dObj
+		}
+	}
+	return res, len(res) > 0
+}
+
+// filterPayload removes the matched paths of provided JSON data.
+func (h *deviceHandler) filterPayload(data interface{}) interface{} {
+	if len(h.payloadFilters) > 0 {
+		if rem, value := h._filterPayload("", data); !rem {
+			return value
+		}
+		return nil
+	}
+	return data
+}
+
+// _filterPayload removes the matched paths of provided JSON data. This function will be called recursively.
+func (h *deviceHandler) _filterPayload(path string, data interface{}) (bool, interface{}) {
+	if mData, ok := data.(map[string]interface{}); ok {
+		for name, value := range mData {
+			rem, val := h._filterPayload(fmt.Sprintf("%s/%s", path, name), value)
+			if rem {
+				delete(mData, name)
+			} else {
+				mData[name] = val
+			}
+		}
+		return len(mData) == 0, data
+	} else if aData, ok := data.([]interface{}); ok {
+		for index := len(aData) - 1; index >= 0; index-- {
+			rem, val := h._filterPayload(fmt.Sprintf("%s/%d", path, index), aData[index])
+			if rem {
+				aData = append(aData[:index], aData[index+1:]...)
+			} else {
+				aData[index] = val
+			}
+		}
+		return len(aData) == 0, aData
+	}
+
+	h.Debug("Found JSON path", map[string]interface{}{"path": path})
+	for _, filter := range h.payloadFilters {
+		if filter.MatchString(path) {
+			h.Debug("Excluded JSON path", map[string]interface{}{"path": path})
+			return true, data
+		}
+	}
+	return false, data
+}
+
+// toShadowMessage convert Ditto message to device shadow message if needed.
+func (h *deviceHandler) toShadowMessages(env *protocol.Envelope) ([]*message.Message, bool) {
+	messages := []*message.Message{}
+	if h.isShadowMessage(env) {
+		if h.topicFilter != nil && h.topicFilter.MatchString(env.Topic.String()) {
+			h.Debug("Excluded message with topic", map[string]interface{}{"topic": env.Topic.String()})
+			return messages, true
+		}
+
+		value := integrate(strings.Split(env.Path, "/"), env.Value)
+		value = h.filterPayload(value)
+		if value == nil {
+			return messages, true
+		}
+
+		// Prepare update messages for found atributes.
+		search(valueAttributesTag, value, func(attributes map[string]interface{}) {
+			message := h.toShadowMessage(env, "", attributes)
+			messages = append(messages, message)
```

Comment:
looks like there is no need to assign new variable here. Same for the feature search below
```suggestion
			messages = append(messages, h.toShadowMessage(env, "", attributes))
```